[Infinite loop] Infinite loop related to noqa

Ruff 33394e4a693077ffe966af772fd5fec8853f786e
```
ruff . --fix
```
on
```
from io import SEEK_CUR, SEEK_END, BytesIO, open  # noqa
class KaitaiStruct:
    @classmethod
    def from_file(cls, filename):
        f = open(filename, "rb")
```
shows
```
error: Failed to converge after 100 iterations.

This indicates a bug in `ruff`. If you could open an issue at:

    https://github.com/charliermarsh/ruff/issues/new?title=%5BInfinite%20loop%5D

...quoting the contents of `Desktop/RunEveryCommand/Ruff/Broken/848376472PY_FILE_TEST_1124330.py`, along with the `pyproject.toml` settings and executed command, we'd be very appreciative!

Desktop/RunEveryCommand/Ruff/Broken/848376472PY_FILE_TEST_1124330.py:1:51: PGH004 Use specific rule codes when using `noqa`
Desktop/RunEveryCommand/Ruff/Broken/848376472PY_FILE_TEST_1124330.py:2:7: D101 Missing docstring in public class
Desktop/RunEveryCommand/Ruff/Broken/848376472PY_FILE_TEST_1124330.py:4:9: ANN206 Missing return type annotation for classmethod `from_file`
Desktop/RunEveryCommand/Ruff/Broken/848376472PY_FILE_TEST_1124330.py:4:9: D102 Missing docstring in public method
Desktop/RunEveryCommand/Ruff/Broken/848376472PY_FILE_TEST_1124330.py:4:19: ANN102 Missing type annotation for `cls` in classmethod
Desktop/RunEveryCommand/Ruff/Broken/848376472PY_FILE_TEST_1124330.py:4:24: ANN001 Missing type annotation for function argument `filename`
Desktop/RunEveryCommand/Ruff/Broken/848376472PY_FILE_TEST_1124330.py:5:9: UP020 Use builtin `open`
Found 107 errors (100 fixed, 7 remaining).

```

--- a/crates/ruff/resources/test/fixtures/pyupgrade/UP020.py
+++ b/crates/ruff/resources/test/fixtures/pyupgrade/UP020.py
@@ -1,9 +1,9 @@
-from io import open
-
-with open("f.txt") as f:
-    print(f.read())
-
 import io
 
 with io.open("f.txt", mode="r", buffering=-1, **kwargs) as f:
+    print(f.read())
+
+from io import open
+
+with open("f.txt") as f:
     print(f.read())

--- a/crates/ruff/src/rules/pyupgrade/mod.rs
+++ b/crates/ruff/src/rules/pyupgrade/mod.rs
@@ -38,6 +38,7 @@ mod tests {
     #[test_case(Rule::RedundantOpenModes, Path::new("UP015.py"); "UP015")]
     #[test_case(Rule::NativeLiterals, Path::new("UP018.py"); "UP018")]
     #[test_case(Rule::TypingTextStrAlias, Path::new("UP019.py"); "UP019")]
+    #[test_case(Rule::OpenAlias, Path::new("UP020.py"); "UP020")]
     #[test_case(Rule::ReplaceUniversalNewlines, Path::new("UP021.py"); "UP021")]
     #[test_case(Rule::ReplaceStdoutStderr, Path::new("UP022.py"); "UP022")]
     #[test_case(Rule::DeprecatedCElementTree, Path::new("UP023.py"); "UP023")]

--- a/crates/ruff/src/rules/pyupgrade/rules/open_alias.rs
+++ b/crates/ruff/src/rules/pyupgrade/rules/open_alias.rs
@@ -1,6 +1,6 @@
 use rustpython_parser::ast::Expr;
 
-use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Fix};
+use ruff_diagnostics::{AutofixKind, Diagnostic, Fix, Violation};
 use ruff_macros::{derive_message_formats, violation};
 use ruff_python_ast::types::Range;
 
@@ -8,16 +8,21 @@ use crate::checkers::ast::Checker;
 use crate::registry::AsRule;
 
 #[violation]
-pub struct OpenAlias;
+pub struct OpenAlias {
+    pub fixable: bool,
+}
 
-impl AlwaysAutofixableViolation for OpenAlias {
+impl Violation for OpenAlias {
+    const AUTOFIX: AutofixKind = AutofixKind::Sometimes;
+
     #[derive_message_formats]
     fn message(&self) -> String {
         format!("Use builtin `open`")
     }
 
-    fn autofix_title(&self) -> String {
-        "Replace with builtin `open`".to_string()
+    fn autofix_title_formatter(&self) -> Option<fn(&Self) -> String> {
+        self.fixable
+            .then_some(|_| format!("Replace with builtin `open`"))
     }
 }
 
@@ -28,8 +33,12 @@ pub fn open_alias(checker: &mut Checker, expr: &Expr, func: &Expr) {
         .resolve_call_path(func)
         .map_or(false, |call_path| call_path.as_slice() == ["io", "open"])
     {
-        let mut diagnostic = Diagnostic::new(OpenAlias, Range::from(expr));
-        if checker.patch(diagnostic.kind.rule()) {
+        let fixable = checker
+            .ctx
+            .find_binding("open")
+            .map_or(true, |binding| binding.kind.is_builtin());
+        let mut diagnostic = Diagnostic::new(OpenAlias { fixable }, Range::from(expr));
+        if fixable && checker.patch(diagnostic.kind.rule()) {
             diagnostic.amend(Fix::replacement(
                 "open".to_string(),
                 func.location,

